fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvRewriteStrategy.java

```diff
@@ -0,0 +1,138 @@
+            if (optimizerContext.getCandidateMvs() == null ||
+                    optimizerContext.getCandidateMvs().isEmpty()) {
+                return false;
+            }
+            if (optimizerConfig.isRuleSetTypeDisable(RuleSetType.SINGLE_TABLE_MV_REWRITE) &&
+                    optimizerConfig.isRuleSetTypeDisable(RuleSetType.MULTI_TABLE_MV_REWRITE)) {
+                return false;
+            }
+            return true;
+        }
+
+        private boolean isEnableRBOSingleViewRewrite() {
+            return optimizerContext.getQueryMaterializationContext() != null
+                    && optimizerContext.getQueryMaterializationContext().getLogicalTreeWithView() != null
+                    && !optimizerConfig.isRuleSetTypeDisable(RuleSetType.SINGLE_TABLE_MV_REWRITE)
+                    && optimizerContext.getCandidateMvs().stream().anyMatch(MaterializationContext::isSingleTable);
+        }
+
+        private boolean isEnableRBOSingleTableRewrite(OptExpression queryPlan) {
+            // if disable single mv rewrite, return false.
+            if (optimizerConfig.isRuleSetTypeDisable(RuleSetType.SINGLE_TABLE_MV_REWRITE)) {
+                return false;
+            }
+            // If query only has one table use single table rewrite, view delta only rewrites multi-tables query.
+            if (!sessionVariable.isEnableMaterializedViewSingleTableViewDeltaRewrite() &&
+                    MvUtils.getAllTables(queryPlan).size() <= 1) {
+                return true;
+            }
+            // If view delta is enabled and there are multi-table mvs, return false.
+            // if mv has multi table sources, we will process it in memo to support view delta join rewrite
+            if (sessionVariable.isEnableMaterializedViewViewDeltaRewrite() &&
+                    optimizerContext.getCandidateMvs().stream().anyMatch(MaterializationContext::hasMultiTables)) {
+                return false;
+            }
+            return true;
+        }
+
+        private boolean isEnableCBOSingleTableRewrite(OptExpression queryPlan) {
+            if (sessionVariable.isEnableMaterializedViewViewDeltaRewrite() &&
+                    optimizerContext.getCandidateMvs().stream().anyMatch(MaterializationContext::hasMultiTables)) {
+                return true;
+            }
+            return false;
+        }
+
+        private boolean isEnableCBOMultiTableRewrite(OptExpression queryPlan) {
+            if (!sessionVariable.isEnableMaterializedViewSingleTableViewDeltaRewrite() &&
+                    MvUtils.getAllTables(queryPlan).size() <= 1) {
+                return false;
+            }
+            return true;
+        }
+    }
+
+    public static MvRewriteStrategy getMvRewriteStrategy(OptimizerContext optimizerContext,
+                                                         ConnectContext connectContext,
+                                                         OptExpression queryPlan) {
```

Comment:
is just using List<Table> or tableSize of queryPlan enough?

Reply:
We can use it for following extends.